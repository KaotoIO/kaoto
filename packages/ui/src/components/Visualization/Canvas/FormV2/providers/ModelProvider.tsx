import { createContext, FunctionComponent, PropsWithChildren, useMemo } from 'react';

export interface ModelContextValue {
  model: unknown;
  errors?: Record<string, string[]>;
  onPropertyChange: (propName: string, value: unknown) => void;
  disabled?: boolean;
}

export const ModelContext = createContext<ModelContextValue>({ model: {}, onPropertyChange: () => {} });

export const ModelContextProvider: FunctionComponent<PropsWithChildren<ModelContextValue>> = ({
  model,
  errors,
  onPropertyChange,
  disabled,
  children,
}) => {
  const value = useMemo(() => {
    return { model, errors, onPropertyChange, disabled };
  }, [model, errors, onPropertyChange, disabled]);

  return <ModelContext.Provider value={value}>{children}</ModelContext.Provider>;
};
