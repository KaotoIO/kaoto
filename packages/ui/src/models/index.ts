export * from './camel-catalog-index';
export * from './camel-components-catalog';
export * from './camel-processors-catalog';
export * from './catalog-kind';
export * from './flow';
export * from './kamelets-catalog';
export * from './schema';
