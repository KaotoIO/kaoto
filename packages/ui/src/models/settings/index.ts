export * from './default-settings-adapter';
export * from './settings.model';
