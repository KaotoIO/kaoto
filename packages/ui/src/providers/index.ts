export * from './catalog-schema-loader.provider';
export * from './entities.provider';
