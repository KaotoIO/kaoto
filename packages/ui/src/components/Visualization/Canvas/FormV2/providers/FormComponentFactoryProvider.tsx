import { createContext, FunctionComponent, PropsWithChildren, useCallback } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';
import { BooleanField } from '../fields/BooleanField';
import { DisabledField } from '../fields/DisabledField';
import { ObjectField } from '../fields/ObjectField/ObjectField';
import { OneOfField } from '../fields/OneOfField/OneOfField';
import { StringField } from '../fields/StringField';
import { PasswordField } from '../fields/PasswordField';
import { FieldProps } from '../typings';

type FormComponentFactoryContextValue = (schema: KaotoSchemaDefinition['schema']) => FunctionComponent<FieldProps>;

export const FormComponentFactoryContext = createContext<FormComponentFactoryContextValue | undefined>(undefined);

export const FormComponentFactoryProvider: FunctionComponent<PropsWithChildren> = ({ children }) => {
  const factory = useCallback<FormComponentFactoryContextValue>((schema) => {
    if (schema.format === 'password') {
      return PasswordField;
    }
    switch (schema.type) {
      case 'string':
      case 'number':
      case 'integer':
        return StringField;
      case 'boolean':
        return BooleanField;
      case 'object':
        return ObjectField;
      case 'array':
        return () => <div>Array field</div>;
    }

    if (Array.isArray(schema.oneOf)) {
      return OneOfField;
    } else if (Array.isArray(schema.anyOf)) {
      return () => <div>AnyOf field</div>;
    }

    return DisabledField;
  }, []);

  return <FormComponentFactoryContext.Provider value={factory}>{children}</FormComponentFactoryContext.Provider>;
};
