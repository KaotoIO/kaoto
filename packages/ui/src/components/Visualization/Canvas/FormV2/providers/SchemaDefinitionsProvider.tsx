import { createContext, FunctionComponent, PropsWithChildren } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';

type SchemaDefinitionsContextValue = Record<string, KaotoSchemaDefinition['schema']>;

export const SchemaDefinitionsContext = createContext<SchemaDefinitionsContextValue>({});

export const SchemaDefinitionsProvider: FunctionComponent<
  PropsWithChildren<{ schema: KaotoSchemaDefinition['schema'] }>
> = ({ schema, children }) => {
  const definitions: SchemaDefinitionsContextValue = schema.definitions ?? {};

  return <SchemaDefinitionsContext.Provider value={definitions}>{children}</SchemaDefinitionsContext.Provider>;
};
