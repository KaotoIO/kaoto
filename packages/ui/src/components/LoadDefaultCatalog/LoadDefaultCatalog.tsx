import {
  Bullseye,
  Button,
  EmptyState,
  EmptyStateActions,
  EmptyStateBody,
  EmptyStateFooter,
  EmptyStateHeader,
  EmptyStateIcon,
  EmptyStateVariant,
  ExpandableSection,
} from '@patternfly/react-core';
import { ExclamationCircleIcon } from '@patternfly/react-icons';
import { FunctionComponent, PropsWithChildren, useContext } from 'react';
import { useReloadContext } from '../../hooks/useReloadContext/useReloadContext';
import { SettingsContext } from '../../providers';

interface ILoadDefaultCatalogProps {
  errorMessage: string;
}

export const LoadDefaultCatalog: FunctionComponent<PropsWithChildren<ILoadDefaultCatalogProps>> = (props) => {
  const settingsAdapter = useContext(SettingsContext);
  const { reloadPage } = useReloadContext();

  const reloadCatalog = () => {
    const currentSettings = settingsAdapter.getSettings();
    settingsAdapter.saveSettings({ ...currentSettings, catalogUrl: '' });
    reloadPage();
  };

  return (
    <Bullseye>
      <EmptyState variant={EmptyStateVariant.lg} data-testid="load-default-catalog">
        <EmptyStateHeader
          titleText="The Catalog couldn't be loaded"
          headingLevel="h4"
          icon={<EmptyStateIcon icon={ExclamationCircleIcon} />}
        />

        <EmptyStateBody>{props.children}</EmptyStateBody>

        <EmptyStateFooter>
          <EmptyStateActions>
            <Button variant="primary" onClick={reloadCatalog}>
              Reload with default Catalog
            </Button>
          </EmptyStateActions>
          <ExpandableSection
            toggleText="Error details"
            toggleId="expandable-section-toggle"
            contentId="expandable-section-content"
          >
            <code>
              <pre>{JSON.stringify(props.errorMessage, null, 2)}</pre>
            </code>
          </ExpandableSection>
        </EmptyStateFooter>
      </EmptyState>
    </Bullseye>
  );
};
