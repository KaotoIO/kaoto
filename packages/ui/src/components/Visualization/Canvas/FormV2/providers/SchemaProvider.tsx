import { createContext, FunctionComponent, PropsWithChildren, useContext, useMemo } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';
import { resolveSchemaWithRef } from '../../../../../utils';
import { SchemaDefinitionsContext } from './SchemaDefinitionsProvider';

export interface SchemaContextValue {
  schema: KaotoSchemaDefinition['schema'];
  definitions: Record<string, KaotoSchemaDefinition['schema']>;
}

export const SchemaContext = createContext<SchemaContextValue>({ schema: {}, definitions: {} });

export const SchemaProvider: FunctionComponent<PropsWithChildren<{ schema: KaotoSchemaDefinition['schema'] }>> = ({
  schema,
  children,
}) => {
  const { definitions, omitFields } = useContext(SchemaDefinitionsContext);

  const value = useMemo(() => {
    const resolvedSchema = resolveSchemaWithRef(schema, definitions);

    if (Array.isArray(resolvedSchema.anyOf)) {
      resolvedSchema.anyOf = resolvedSchema.anyOf.map((anyOfSchema) => resolveSchemaWithRef(anyOfSchema, definitions));
    }
    if (Array.isArray(resolvedSchema.oneOf)) {
      resolvedSchema.oneOf = resolvedSchema.oneOf.map((oneOfSchema) => resolveSchemaWithRef(oneOfSchema, definitions));
    }

    omitFields.forEach((field) => {
      delete resolvedSchema.properties?.[field];
    });

    return { schema: resolvedSchema, definitions };
  }, [definitions, omitFields, schema]);

  return <SchemaContext.Provider value={value}>{children}</SchemaContext.Provider>;
};
