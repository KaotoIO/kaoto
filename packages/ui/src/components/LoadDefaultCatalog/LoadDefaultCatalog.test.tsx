import { render } from '@testing-library/react';
import { AbstractSettingsAdapter, DefaultSettingsAdapter } from '../../models/settings';
import { ReloadContext, SettingsProvider } from '../../providers';
import { LoadDefaultCatalog } from './LoadDefaultCatalog';

describe('LoadDefaultCatalog', () => {
  let reloadPage: jest.Mock;
  let settingsAdapter: AbstractSettingsAdapter;

  beforeEach(() => {
    reloadPage = jest.fn();
    settingsAdapter = new DefaultSettingsAdapter();
  });

  it('should render correctly', () => {
    const wrapper = render(
      <ReloadContext.Provider value={{ reloadPage, lastRender: 0 }}>
        <SettingsProvider adapter={settingsAdapter}>
          <LoadDefaultCatalog errorMessage="Test error message" />
        </SettingsProvider>
      </ReloadContext.Provider>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render child components', () => {
    const wrapper = render(
      <ReloadContext.Provider value={{ reloadPage, lastRender: 0 }}>
        <SettingsProvider adapter={settingsAdapter}>
          <LoadDefaultCatalog errorMessage="Test error message">
            <div>Test children</div>
          </LoadDefaultCatalog>
        </SettingsProvider>
      </ReloadContext.Provider>,
    );

    const child = wrapper.getByText('Test children');

    expect(child).toBeInTheDocument();
  });

  it('should call reloadCatalog on button click', () => {
    const wrapper = render(
      <ReloadContext.Provider value={{ reloadPage, lastRender: 0 }}>
        <SettingsProvider adapter={settingsAdapter}>
          <LoadDefaultCatalog errorMessage="Test error message">
            <div>Test children</div>
          </LoadDefaultCatalog>
        </SettingsProvider>
      </ReloadContext.Provider>,
    );

    const button = wrapper.getByText('Reload with default Catalog');
    button.click();

    expect(reloadPage).toHaveBeenCalledTimes(1);
  });
});
