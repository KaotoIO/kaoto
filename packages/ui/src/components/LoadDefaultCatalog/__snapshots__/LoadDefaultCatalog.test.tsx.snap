// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadDefaultCatalog should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="simple-zustand-devtools-Schemas Store"
    />
    <div>
      <div
        class="pf-v5-l-bullseye"
      >
        <div
          class="pf-v5-c-empty-state pf-m-lg"
          data-testid="load-default-catalog"
        >
          <div
            class="pf-v5-c-empty-state__content"
          >
            <div
              class="pf-v5-c-empty-state__header"
            >
              <div
                class="pf-v5-c-empty-state__icon"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                  />
                </svg>
              </div>
              <div
                class="pf-v5-c-empty-state__title"
              >
                <h4
                  class="pf-v5-c-empty-state__title-text"
                >
                  The Catalog couldn't be loaded
                </h4>
              </div>
            </div>
            <div
              class="pf-v5-c-empty-state__body"
            />
            <div
              class="pf-v5-c-empty-state__footer"
            >
              <div
                class="pf-v5-c-empty-state__actions"
              >
                <button
                  aria-disabled="false"
                  class="pf-v5-c-button pf-m-primary"
                  data-ouia-component-id="OUIA-Generated-Button-primary-1"
                  data-ouia-component-type="PF5/Button"
                  data-ouia-safe="true"
                  type="button"
                >
                  Reload with default Catalog
                </button>
              </div>
              <div
                class="pf-v5-c-expandable-section"
              >
                <button
                  aria-controls="expandable-section-content"
                  aria-expanded="false"
                  class="pf-v5-c-expandable-section__toggle"
                  id="expandable-section-toggle"
                  type="button"
                >
                  <span
                    class="pf-v5-c-expandable-section__toggle-icon"
                  >
                    <svg
                      aria-hidden="true"
                      class="pf-v5-svg"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      viewBox="0 0 256 512"
                      width="1em"
                    >
                      <path
                        d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                      />
                    </svg>
                  </span>
                  <span
                    class="pf-v5-c-expandable-section__toggle-text"
                  >
                    Error details
                  </span>
                </button>
                <div
                  aria-labelledby="expandable-section-toggle"
                  class="pf-v5-c-expandable-section__content"
                  hidden=""
                  id="expandable-section-content"
                  role="region"
                >
                  <code>
                    <pre>
                      "Test error message"
                    </pre>
                  </code>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="pf-v5-l-bullseye"
    >
      <div
        class="pf-v5-c-empty-state pf-m-lg"
        data-testid="load-default-catalog"
      >
        <div
          class="pf-v5-c-empty-state__content"
        >
          <div
            class="pf-v5-c-empty-state__header"
          >
            <div
              class="pf-v5-c-empty-state__icon"
            >
              <svg
                aria-hidden="true"
                class="pf-v5-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                />
              </svg>
            </div>
            <div
              class="pf-v5-c-empty-state__title"
            >
              <h4
                class="pf-v5-c-empty-state__title-text"
              >
                The Catalog couldn't be loaded
              </h4>
            </div>
          </div>
          <div
            class="pf-v5-c-empty-state__body"
          />
          <div
            class="pf-v5-c-empty-state__footer"
          >
            <div
              class="pf-v5-c-empty-state__actions"
            >
              <button
                aria-disabled="false"
                class="pf-v5-c-button pf-m-primary"
                data-ouia-component-id="OUIA-Generated-Button-primary-1"
                data-ouia-component-type="PF5/Button"
                data-ouia-safe="true"
                type="button"
              >
                Reload with default Catalog
              </button>
            </div>
            <div
              class="pf-v5-c-expandable-section"
            >
              <button
                aria-controls="expandable-section-content"
                aria-expanded="false"
                class="pf-v5-c-expandable-section__toggle"
                id="expandable-section-toggle"
                type="button"
              >
                <span
                  class="pf-v5-c-expandable-section__toggle-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 256 512"
                    width="1em"
                  >
                    <path
                      d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                    />
                  </svg>
                </span>
                <span
                  class="pf-v5-c-expandable-section__toggle-text"
                >
                  Error details
                </span>
              </button>
              <div
                aria-labelledby="expandable-section-toggle"
                class="pf-v5-c-expandable-section__content"
                hidden=""
                id="expandable-section-content"
                role="region"
              >
                <code>
                  <pre>
                    "Test error message"
                  </pre>
                </code>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
